### PR TITLE
Feature/enable replication graph

### PR DESCRIPTION
**Pull request by m-samiec**
#### Description
Replication graph support added.

Engine 23 PR : https://github.com/improbableio/UnrealEngine/pull/376
Engine 24 PR : https://github.com/improbableio/UnrealEngine/pull/384
TestGyms PR: https://github.com/spatialos/UnrealGDKTestGyms/pull/50

#### Release note
Feature: Replication graph support has been added. An example of functionality can be seen in the TestGyms project.

#### Tests
Tested in TestGyms PR

#### Documentation
Follow up - https://improbableio.atlassian.net/browse/UNR-3092

#### Primary reviewers
@improbable-valentyn @mironec

### CODE REVIEW

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 1]
```diff
@@ -1515,6 +1522,13 @@ int32 USpatialNetDriver::ServerReplicateActors(float DeltaSeconds)
 		{
 			new(ConnectionViewers)FNetViewer(ClientConnection, DeltaSeconds);
 
+			// send ClientAdjustment if necessary
+			// we do this here so that we send a maximum of one per packet to that client; there is no value in stacking additional corrections
+			if (ClientConnection->PlayerController)
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]:
```suggestion
			if (ClientConnection->PlayerController != nullptr)
```

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialReplicationGraph.cpp [hunk 2]
```diff
@@ -0,0 +1,22 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "EngineClasses/SpatialReplicationGraph.h"
+
+#include "EngineClasses/SpatialActorChannel.h"
+#include "EngineClasses/SpatialNetDriver.h"
+
+
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialReplicationGraph.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 3]
```diff
@@ -1503,6 +1504,12 @@ void USpatialReceiver::OnComponentUpdate(const Worker_ComponentUpdateOp& Op)
 			if (AActor* Actor = Cast<AActor>(PackageMap->GetObjectFromEntityId(Op.entity_id)))
 			{
 				Channel = RecreateDormantSpatialChannel(Actor, Op.entity_id);
+
+				// As we haven't removed the dormant component just yet, this might be a single replication update where the actor
+				// remains dormant. Add it back to pending dormancy so the local worker can clean up the channel. If we do process
+				// a dormant component removal later in this frame, we'll clear the channel from pending dormancy channel then.
+				NetDriver->AddPendingDormantChannel(Channel);
+
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h [hunk 4]
```diff
@@ -0,0 +1,25 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+
```
Comment (improbable-valentyn) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h]:
```suggestion
```

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs [hunk 5]
```diff
@@ -40,6 +40,7 @@ public SpatialGDK(ReadOnlyTargetRules Target) : base(Target)
                 "OnlineSubsystemUtils",
                 "InputCore",
                 "Sockets",
+                "ReplicationGraph"
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: Slightly worried that we now have a dependency on ReplicationGraph, even though it is not strictly required for us to function, but I guess the solution would be to create a SpatialReplicationGraph module which seems like a lot of unnecessary work.

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp [hunk 6]
```diff
@@ -526,12 +531,6 @@ int64 USpatialActorChannel::ReplicateActor()
 #if USE_NETWORK_PROFILER 
 	const uint32 ActorReplicateStartTime = GNetworkProfiler.IsTrackingEnabled() ? FPlatformTime::Cycles() : 0;
 #endif
-	// Epic does this at the net driver level, per connection. See UNetDriver::ServerReplicateActors().
-	// However, we have many player controllers sharing one connection, so we do it at the actor level before replication.
-	if (APlayerController* PlayerController = Cast<APlayerController>(Actor))
```
Comment (improbable-danny) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialActorChannel.cpp]: Is this still required for non-replication graph?

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs (hunk 5 above)
Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: Yeah I'm not so worried, it's a core runtime module, and only has a handful of files. Also, most spatial games should be making use of replication graphs going forward.

File: SpatialGDK/SpatialGDK.uplugin [hunk 7]
```diff
@@ -58,7 +58,7 @@
       "Enabled": true
     },
     {
-      "Name": "GameplayAbilities",
+      "Name": "ReplicationGraph",
```
Comment (aleximprobable) [inline SpatialGDK/SpatialGDK.uplugin]: Is it on purpose, that `GameplayAbilities` is removed here?

Comment (m-samiec) [reply SpatialGDK/SpatialGDK.uplugin]: Yep - should have mentioned, this is a driveby - it was a no-longer required dependency.

File: SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h [hunk 8]
```diff
@@ -0,0 +1,24 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#pragma once
+
+#include "CoreMinimal.h"
+#include "ReplicationGraph.h"
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Public/EngineClasses/SpatialReplicationGraph.h]: nit: "ReplicationGraph.h" should be included before #include "CoreMinimal.h"

File: SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs (hunk 5 above)
Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/SpatialGDK.Build.cs]: We may also make it a conditional dependency if needed?

File: SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp [hunk 9]
```diff
@@ -1462,6 +1464,11 @@ int32 USpatialNetDriver::ServerReplicateActors(float DeltaSeconds)
 	check(SpatialConnection && SpatialConnection->bReliableSpatialConnection);
 	check(World);
 
+	if (UReplicationDriver* RepDriver = GetReplicationDriver())
```
Comment (aleximprobable) [inline SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Could we call `RepDriver->ServerReplicateActors(DeltaSeconds);` conditionally in `void USpatialNetDriver::TickFlush(float DeltaTime)` instead?

Comment (aleximprobable) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: It's there to resemble UE4 code, resolving.

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/EngineClasses/SpatialNetDriver.cpp]: Followed ordering in `NetworkDriver.cpp` - will move the check

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 10]
```diff
@@ -808,10 +809,13 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
 	}
 
 	// Set up actor channel.
-	USpatialActorChannel* Channel = Cast<USpatialActorChannel>(Connection->CreateChannelByName(NAME_Actor, NetDriver->IsServer() ? EChannelCreateFlags::OpenedLocally : EChannelCreateFlags::None));
-	Channel->RefreshAuthority();
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Completely missed this when doing my PR, should have been in the block below.

File: SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp [hunk 11]
```diff
@@ -825,11 +829,16 @@ void USpatialReceiver::ReceiveActor(Worker_EntityId EntityId)
...
 #if ENGINE_MINOR_VERSION <= 22
-	Channel->SetChannelActor(EntityActor);
+		Channel->SetChannelActor(EntityActor);
 #else
-	Channel->SetChannelActor(EntityActor, ESetChannelActorFlags::None);
+		Channel->SetChannelActor(EntityActor, ESetChannelActorFlags::None);
 #endif
+	}
+  
+  Channel->RefreshAuthority();
```
Comment (mironec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: I think one tab out?

Comment (m-samiec) [inline SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]:
```suggestion
	Channel->RefreshAuthority();
```

Comment (m-samiec) [reply SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp]: Fixed

**Pull request merged**